reactor-core/src/main/java/reactor/core/publisher/Traces.java

```diff
@@ -97,80 +95,164 @@ static String extractOperatorAssemblyInformation(String source) {
-			apiLine = "";
-			userCodeLine = traces.get(0);
+		while (traces.hasNext()) {
+			prevLine = currentLine;
+			currentLine = traces.next();
+
+			if (currentLine.isUserCode()) {
+				// Currently on user code line, previous one is API. Attempt to create something in the form
+				// "Flux.map ⇢ user.code.Class.method(Class.java:123)".
+				return new String[]{
+					prevLine.withoutPublisherPackagePrefix().withoutLocationSuffix().toString(),
+					"at " + currentLine};
+			}
 		}
-		else if (i == traces.size()) {
-			//we skipped ALL lines, meaning they're all reactor API lines. We'll fully display the last one
-			apiLine = "";
-			userCodeLine = traces.get(i-1).replaceFirst("reactor.core.publisher.", "");
+
+		// We skipped ALL lines, meaning they're all Reactor API lines. We'll fully display the last
+		// one.
+		return new String[]{currentLine.withoutPublisherPackagePrefix().toString()};
+	}
+
+	/**
+	 * Returns an iterator over all trimmed non-empty lines in the given source string.
+	 *
+	 * @implNote This implementation attempts to minimize allocations.
+	 */
+	private static Iterator<Substring> trimmedNonemptyLines(String source) {
+		return new Iterator<Substring>() {
+			private int index = 0;
+			@Nullable
+			private Substring next = getNextLine();
+
+			@Override
+			public boolean hasNext() {
+				return next != null;
+			}
+
+			@Override
+			public Substring next() {
+				Substring current = next;
+				if (current == null) {
+					throw new NoSuchElementException();
+				}
+				next = getNextLine();
+				return current;
+			}
+
+			@Nullable
+			private Substring getNextLine() {
+				while (index < source.length()) {
+					int end = source.indexOf('\n', index);
+					if (end == -1) {
+						end = source.length();
+					}
+					Substring line = new Substring(source, index, end).trim();
+					index = end + 1;
+					if (!line.isEmpty()) {
+						return line;
+					}
+				}
+				return null;
+			}
+		};
+	}
+
+	// XXX: Explain.
+	private static final class Substring {
+		private final String str;
+		private final int start;
+		private final int end;
+
+		Substring(String str, int start, int end) {
+			this.str = str;
+			this.start = start;
+			this.end = end;
 		}
-		else {
-			//currently on user code line, previous one is API
-			apiLine = traces.get(i - 1);
-			userCodeLine = traces.get(i);
+
+		Substring trim() {
+			int newStart = start;
+			while (newStart < end && str.charAt(newStart) <= ' ') {
+				newStart++;
+			}
+			int newEnd = end;
+			while (newEnd > newStart && str.charAt(newEnd - 1) <= ' ') {
+				newEnd--;
+			}
+			return newStart == start && newEnd == end ? this : new Substring(str, newStart, newEnd);
 		}
 
-		//now we want something in the form "Flux.map ⇢ user.code.Class.method(Class.java:123)"
-		if (apiLine.isEmpty()) return new String[] { userCodeLine };
+		boolean isEmpty() {
+			return start == end;
+		}
+
+		boolean startsWith(String prefix) {
+			return str.startsWith(prefix, start);
```

Comment:
For our specific use case it probably won't be a problem but in general this is incomplete as the `end` index is not considered. Perhaps a simple check for `start + prefix.length() < end` is also required to make it correct?
With that, I'd argue a bunch of unit tests for this inner class would be helpful. It can be made package private and we can test it in `TracesTest.java`.